app/lib/shared/search_service.dart

```diff
@@ -0,0 +1,156 @@
+const String searchIndexNotReadyText = 'Not ready yet.';
+
+/// Package search index and lookup.
+abstract class PackageIndex {
+  bool get isReady;
+  Future<int> indexSize();
+  Future add(PackageDocument doc);
+  Future addAll(Iterable<PackageDocument> documents);
+  Future contains(String url, String version, String devVersion);
+  Future removeUrl(String url);
+  Future merge();
+  Future<PackageSearchResult> search(PackageQuery query);
+}
+
+/// A summary information about a package that goes into the search index.
+///
+/// It is also part of the data structure returned by a search query, except for
+/// the [readme] and [popularity] fields, which are excluded when returning the
+/// results.
+class PackageDocument {
+  final String url;
+  final String package;
+  final String version;
+  final String devVersion;
+  final String description;
+  final String lastUpdated;
+  final String readme;
+
+  final List<String> detectedTypes;
+  final double popularity;
+
+  PackageDocument({
+    this.url,
+    this.package,
+    this.version,
+    this.devVersion,
+    this.description,
+    this.lastUpdated,
+    this.readme,
+    this.detectedTypes,
+    this.popularity,
+  });
+
+  factory PackageDocument.parseJson(Map json) => new PackageDocument(
+        url: json['url'],
+        package: json['package'],
+        version: json['version'],
+        devVersion: json['devVersion'],
+        description: json['description'],
+        lastUpdated: json['lastUpdated'],
+        readme: json['readme'],
+        detectedTypes: json['detectedTypes'],
+        popularity: json['popularity'],
+      );
+
+  Map toJson() => {
+        'url': url,
+        'package': package,
+        'version': version,
+        'devVersion': devVersion,
+        'description': description,
+        'lastUpdated': lastUpdated,
+        'readme': readme,
+        'detectedTypes': detectedTypes,
+        'popularity': popularity,
+      };
+}
+
+class PackageQuery {
+  final String text;
+  final String type;
+  final int offset;
+  final int limit;
+
+  PackageQuery(this.text, {this.type, this.offset, this.limit});
+
+  factory PackageQuery.fromServiceQueryParameters(Map<String, String> params) {
+    final String text = params['q'];
+    String type = params['type'];
+    if (type != null && type.isEmpty) type = null;
+    int offset = int.parse(params['offset'] ?? '0', onError: (_) => 0);
+    int limit =
+        int.parse(params['limit'] ?? '0', onError: (_) => defaultSearchLimit);
+
+    offset = min(maxSearchResults - minSearchLimit, offset);
+    offset = max(0, offset);
+    limit = max(minSearchLimit, limit);
+
+    return new PackageQuery(text, type: type, offset: offset, limit: limit);
+  }
+
+  Map<String, String> toServiceQueryParameters() => {
+        'q': text,
+        'type': type,
+        'offset': offset?.toString(),
+        'limit': limit?.toString(),
+      };
+}
+
+class PackageSearchResult {
+  final String lastUpdated;
```

Comment:
Maybe add a comment that this `lastUpdated` means actually the last run of the indexing system (as opposed to the `PackageDocument.lastUpdated`, which should just be the last time the package/version was indexed, which can be much longer ago, if there were no new "pub publish"es since then)